server/quoridor/quoridor.go

```diff
@@ -0,0 +1,444 @@
+	for _, player := range game.Players {
+		if player.PlayerId == id {
+			return 0, errors.New(fmt.Sprintf("player with id %s alreayd in this game", id.String()))
+		}
+	}
+	barriersForPlayer := 10
+	if len(game.Players) >= 2 {
+		barriersForPlayer = 5
+	}
+	// For each possible player
+	for playerNumber := PlayerOne; playerNumber <= PlayerFour; playerNumber++ {
+		p, present := game.Players[playerNumber]
+		if present {
+			// Make sure they have the correct number of barriers
+			p.Barriers = barriersForPlayer
+			game.Players[playerNumber] = p
+		} else {
+			playerPawn := Piece{
+				Position: startingPositions[playerNumber],
+				Owner:    playerNumber,
+				Type:     Pawn,
+			}
+			// Create a new player with barrier count, starting position, etc.
+			game.Players[playerNumber] = &Player{
+				Barriers:   barriersForPlayer,
+				PlayerId:   id,
+				PlayerName: name,
+				Pawn:       playerPawn,
+			}
+			// Add pawn to board
+			game.Board[playerPawn.Position] = playerPawn
+			return playerNumber, nil
+		}
+	}
+	return -1, errors.New("no open player positions in this game")
+}
+
+// Advances the turn to the next player.
+func (game *Game) NextTurn() {
+	next := int(game.CurrentTurn+1) % len(game.Players)
+	game.CurrentTurn = PlayerPosition(next)
+}
+
+func (game *Game) IsGameOver() bool {
+	return !game.EndDate.IsZero()
+}
+
+func (game *Game) MaybeReturnWinnerPlayerPosition() PlayerPosition {
+	for position, player := range game.Players {
+		winningPosition := winningPositions[position]
+		if player.Pawn.Position.Y == winningPosition.Y ||
+			player.Pawn.Position.X == winningPosition.X {
+			return position
+		}
+	}
+	return -1
+}
+
+// Deep copy of a game. Will copy all Players, Pieces, and Board.
+func (game *Game) Copy() Game {
+	newGame := Game{
+		Id:          game.Id,
+		CurrentTurn: game.CurrentTurn,
+		StartDate:   game.StartDate,
+		EndDate:     game.EndDate,
+		Winner:      game.Winner,
+		Name:        game.Name,
+	}
+
+	newGame.Board = make(Board)
+	for pos, piece := range game.Board {
+		newPiece := piece
+		newGame.Board[pos] = newPiece
+	}
+
+	newGame.Players = make(map[PlayerPosition]*Player)
+	for pos, player := range game.Players {
+		newGame.Players[pos] = &Player{
+			Barriers:   player.Barriers,
+			Pawn:       player.Pawn,
+			PlayerId:   player.PlayerId,
+			PlayerName: player.PlayerName,
+		}
+	}
+	return newGame
+}
+
+// Starts a game by setting the StartDate to the current instant of time. Returns an error if there aren't enough
+// players, or the game has already started.
+func (game *Game) StartGame() error {
+	if !(len(game.Players) == 2 || len(game.Players) == 4) {
+		return errors.New(fmt.Sprintf("can't start game, wrong number of players (%d)", len(game.Players)))
+	}
+	if !game.StartDate.IsZero() {
+		return errors.New(fmt.Sprintf("game already started"))
+	}
+	game.StartDate = time.Now()
+	return nil
+}
+
+// Initialize with default values
+func NewGame(id uuid.UUID, name string) (*Game, error) {
```

Comment:
why would you pass the id in here? Why not just let the "newGame method" make a new one?

Reply:
I find this design more testable - I have complete control over everything the Game needs to set-up. Everything else are just empty default values. 

In testing, I can assert that the game's Id is the one I constructed for the test, rather than just assert that the game has an Id.